### PR TITLE
fix(sdk/compiler): v2 compat - fix mlpipeline-ui-metadata artifact. Part of #5666

### PR DESCRIPTION
**Pull request by Bobgy**
**Description of your changes:**
Part of #5666 
image
The pipeline properly generates the `mlpipeline-ui-metadata` artifact.

Content in the artifact is not fully correct, but the problem is unrelated to this PR, it's tracked in https://github.com/kubeflow/pipelines/issues/5830

```
$ gsutil cat gs://gongyuan-dev/v2-sample-test/data/samples_config-loop-item/pipeline-tensorboard-minio/pipeline-tensorboard-minio-spt2m/create-tensorboard-visualization/mlpipeline-ui-metadata

    {
      "outputs" : [{
        "type": "tensorboard",
        "source": "s3://mlpipeline/tensorboard/logs/489a0259-33f4-4d1a-845f-9b68f388c8c7",
        "image": "gcr.io/deeplearning-platform-release/tf2-cpu.2-3:latest",
        "pod_template_spec": {\"spec\": {\"containers\": [{\"env\": [{\"name\": \"AWS_ACCESS_KEY_ID\", \"valueFrom\": {\"secretKeyRef\": {\"name\": \"mlpipeline-minio-artifact\", \"key\": \"accesskey\"}}}, {\"name\": \"AWS_SECRET_ACCESS_KEY\", \"valueFrom\": {\"secretKeyRef\": {\"name\": \"mlpipeline-minio-artifact\", \"key\": \"secretkey\"}}}, {\"name\": \"AWS_REGION\", \"value\": \"minio\"}, {\"name\": \"S3_ENDPOINT\", \"value\": \"minio-service:9000\"}, {\"name\": \"S3_USE_HTTPS\", \"value\": \"0\"}, {\"name\": \"S3_VERIFY_SSL\", \"value\": \"0\"}]}]}}
      }]
    }
```

**Checklist:**
- [ ] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: sdk/python/kfp/compiler/v2_compat.py [hunk 1]
```diff
@@ -126,13 +126,18 @@ def update_op(op: dsl.ContainerOp,
 
   for artifact_name, spec in sorted(
       component_spec.output_definitions.artifacts.items()):
+    metadata_path = op.file_outputs.get(artifact_name)
+    if not metadata_path:
+        metadata_path = op.output_artifact_paths.get(artifact_name)
+        if not metadata_path:
+            raise Exception(f"Compiler internal error: cannot find output artifact path for artifact '{artifact_name}' in op '{op.name}'")
```
Comment (Bobgy) [inline sdk/python/kfp/compiler/v2_compat.py:133]: The reason I find artifacts in `op.output_artifact_paths` is because `mlpipeline-ui-metadata` artifact has special handling in https://github.com/kubeflow/pipelines/blob/b3c2351a02231f02ecb3f144d3bade06d854ad18/sdk/python/kfp/dsl/_container_op.py#L1206-L1215.

I'm not sure we are ready to remove that handling. Also, when ContainerOps are constructed, we do not know whether it's in v2 compatible mode or v1, therefore the safest fix is to get the artifact here. It seems that this works out at the end, but I'd hope someone knows more about the compiler to double check this.

File: sdk/python/kfp/compiler/v2_compat.py [hunk 2]
```diff
@@ -126,13 +126,18 @@ def update_op(op: dsl.ContainerOp,
 
   for artifact_name, spec in sorted(
       component_spec.output_definitions.artifacts.items()):
+    metadata_path = op.file_outputs.get(artifact_name)
```
Comment (chensun) [inline sdk/python/kfp/compiler/v2_compat.py:129]: This is rather hacky and should be temporary only. 
Maybe leave a TODO here and point to the issue? This should be reverted once we fix the sanitization.

File: samples/core/visualization/tensorboard_gcs.py [hunk 3]
```diff
@@ -70,4 +70,4 @@ def my_pipeline(
     log_dir=f'gs://{{kfp-default-bucket}}/tensorboard/logs/{dsl.RUN_ID_PLACEHOLDER}'
 ):
     prepare_tb_task = prepare_tensorboard(log_dir_uri=log_dir)
-    tensorboard_task = train_op(log_dir=prepare_tb_task.outputs['log_dir_uri'],)
+    tensorboard_task = train_op(log_dir=log_dir).after(prepare_tb_task)
```
Comment (chensun) [inline samples/core/visualization/tensorboard_gcs.py:73]: Also not ideal. so maybe add a comment above to revert this change once we fixed the sanitization issue?

Comment (Bobgy) [reply samples/core/visualization/tensorboard_gcs.py:73]: Maybe I missed anything, what's the problem with this line?

File: sdk/python/kfp/compiler/v2_compat.py (hunk 1 above)
Comment (Bobgy) [reply sdk/python/kfp/compiler/v2_compat.py:133]: Ohh, thanks for letting me know about `kfp.COMPILING_FOR_V2`. I can try using it instead to remove the hack here.

Comment (Bobgy) [reply sdk/python/kfp/compiler/v2_compat.py]: Updated code using `kfp.COMPILING_FOR_V2`, now it's no longer hacky : )

File: samples/core/visualization/tensorboard_gcs.py (hunk 3 above)
Comment (chensun) [reply samples/core/visualization/tensorboard_gcs.py:73]: The version before the change is clear and explicit dependency -- The output from one task is consumed by the second task.
After the change, this dependency becomes indirect. Technically speaking, all artifact can be passed this way. But I don't think this is a recommended pattern. My concern is, without knowing the reason why this is written this way, user may start copy this pattern.

Comment (Bobgy) [reply samples/core/visualization/tensorboard_gcs.py:73]: Note the prepare_tensorboard task does not write anything to the URI. (Previous logic is that it gets an URI as input and output the same URI, which is unnecessary anyway)

Comment (chensun) [reply samples/core/visualization/tensorboard_gcs.py:73]: I see. Thanks for the explanation.

File: components/tensorflow/tensorboard/prepare_tensorboard/component.yaml [hunk 4]
```diff
@@ -39,7 +35,6 @@ implementation:
           }
       ' >"$output_metadata_path"
     - {inputValue: Log dir URI}
-    - {outputPath: Log dir URI}
```
Comment (Ark-kun) [inline components/tensorflow/tensorboard/prepare_tensorboard/component.yaml:42]: This is a breaking change. This output is supposed to be used by most pipelines that use this component. The trainer's Log dir URI should come from the Prepare Tensorboard component.

Comment (Bobgy) [reply components/tensorflow/tensorboard/prepare_tensorboard/component.yaml:42]: Yes, we can add the output back once fixing https://github.com/kubeflow/pipelines/issues/5831.
